### PR TITLE
Allow the use of multiple status banners

### PR DESCRIPTION
**Pull request by jill-cardamon**
### Description
This PR allows the use of multiple status banners by refactoring `StatusView.swift` to manage the display statuses.
- [x] Create `Status` struct to represent a status, store them in a `statuses` array, and refactor as necessary
- [x] Refactor how statuses are shown and hidden by priority
- [x] How will we deal with special values associated with a `Status`? (e.g.: simulation status banner is interactive, allows for changes in simulation speed)
- [x] Fix [#2723 ](https://github.com/mapbox/mapbox-navigation-ios/issues/2723) (hide "enable precise location" banner when precise location is switched on)
- [x] Test (StatusViewTests.swift)
- [x] Update Changelog

### CODE REVIEW

**Comment by 1ec5** [comment]
CocoaPods integration tests are [failing](https://app.circleci.com/pipelines/github/mapbox/mapbox-navigation-ios/2244/workflows/0e0be1ce-8a3c-49e8-be36-3f83472b1623/jobs/24337):

```
[!] CocoaPods could not find compatible versions for pod "Mapbox-iOS-SDK":
  In snapshot (Podfile.lock):
    Mapbox-iOS-SDK (= 6.2.2, ~> 6.0)
```

CocoaPods picked up the edit to v6.3.0 in https://github.com/mapbox/mapbox-gl-native-ios/pull/549#issuecomment-733490715 but apparently not the earlier edit to v6.2.2 in https://github.com/mapbox/mapbox-gl-native-ios/pull/549#issuecomment-733270439:

```
  CDN: trunk Relative path: Specs/a/5/9/Mapbox-iOS-SDK/6.3.0/Mapbox-iOS-SDK.podspec.json modified during this run! Returning local
  CDN: trunk Relative path downloaded: all_pods_versions_3_5_2.txt, save ETag: W/"5fbd8072-29e"
```

This is basically CocoaPods/CocoaPods#10229. We could work around it by forcing the CI job to start with a fresh cache, but at this point, it might be easier to update the integration test fixture by running these commands:

https://github.com/mapbox/mapbox-navigation-ios/blob/72d78457c3e23f854672ce1f37f99cd706bc90c3/scripts/update-version.sh#L35-L37

**Review by Rmkasun-8 — APPROVED** [review]

**Comment by MaximAlien** [comment]
Is this PR ready to be reviewed (I can see that it's marked as `Draft`)? Also it'd be good to fix outstanding conflicts with `main` branch.

**Comment by jill-cardamon** [comment]
Yes. Ready to be reviewed. I have two questions if you could answer them in your review:
- Do I delete method calls in other functions and add a deprecated flag to the definition only?
- Advice on timer related issue: used a dispatch queue to make things happen when I want them to, but when I was testing, I realized that hiding (which uses a dispatch queue) takes a while. Not sure if it’s because of the loop I implemented or something else. But it takes a while for it to disappear (longer than the delay duration). I did not notice this issue in the example with the current statuses we use.

File: MapboxNavigation/NavigationViewController.swift
Comment (ShanMa1991) [inline MapboxNavigation/NavigationViewController.swift:746]: Nit: removing extra whitespaces and cleaning the code would be great.

File: MapboxNavigation/StatusView.swift
Comment (ShanMa1991) [inline MapboxNavigation/StatusView.swift:143]: Nit: removing the unnecessary comment would be better.

File: MapboxNavigation/StatusView.swift
Comment (MaximAlien) [inline MapboxNavigation/StatusView.swift:135]: It'd be better to extend existing APIs instead of replacing existing ones. If you feel that newly added API is more concise previous one can be deprecated/or left without any changes.

File: MapboxNavigation/StatusView.swift
Comment (MaximAlien) [inline MapboxNavigation/StatusView.swift:60]: It'd be better to use enum for status priority.

File: MapboxNavigation/NavigationComponent.swift
Comment (MaximAlien) [inline MapboxNavigation/NavigationComponent.swift:40]: I think such comment would be accurate for `func addNewStatus(status: StatusView.Status)`. I think it'd better to comment with `- parameter` statements in this case. Please also add docs for `addNewStatus(status:)` and `hideStatus(status:)`.

File: MapboxNavigation/NavigationComponent.swift
Comment (MaximAlien) [inline MapboxNavigation/NavigationComponent.swift:44]: I'd also rename `addNewStatus(status:)` to `func show(_ status: StatusView.Status)` and `func hideStatus(using: StatusView.Status)` to `hide(_ status: StatusView.Status)`.

File: MapboxNavigation/TopBannerViewController.swift
Comment (MaximAlien) [inline MapboxNavigation/TopBannerViewController.swift:419]: Is commented part needed? If not, please remove it.

File: MapboxNavigation/StatusView.swift
Comment (1ec5) [reply MapboxNavigation/StatusView.swift:60]: I think the idea was to allow for arbitrary priorities, similar to constraint priorities. But making the type public would make sense.

File: CHANGELOG.md
Comment (1ec5) [inline CHANGELOG.md:9]: Let’s focus on the API change as it affects developers and remove the implementation details:

```suggestion
* `NavigationViewController` can now manage multiple status banners one after another. Renamed the `NavigationViewController.showStatus(title:spinner:duration:animated:interactive:)` method to `NavigationViewController.show(_:)` and added a corresponding `NavigationViewController.hide(_:)` method. Renamed the `NavigationStatusPresenter.showStatus(title:spinner:duration:animated:interactive:)` method to `NavigationStatusPresenter.show(_:)` and added a `NavigationStatusPresenter.hide(_:)` method. ([#2747](https://github.com/mapbox/mapbox-navigation-ios/pull/2747))
```

File: Sources/MapboxNavigation/NavigationComponent.swift
Comment (1ec5) [inline Sources/MapboxNavigation/NavigationComponent.swift:42]: This is technically a backwards-incompatible change, because an application may expect to be able to call `showStatus(title:spinner:duration:animated:interactive:)`. In order for these changes to go into v1._x_, the old method needs to be left in place but marked as deprecated. Its implementation inside StatusView can create a Status and pass it into the new `show(_:)` method.

File: Sources/MapboxNavigation/NavigationViewController.swift
Comment (1ec5) [inline Sources/MapboxNavigation/NavigationViewController.swift:471]: Similarly, this is a backwards-incompatible change, so `showStatus(title:spinner:duration:animated:interactive:)` will have to be brought back as a deprecated method.

File: Tests/MapboxNavigationTests/StatusViewTests.swift
Comment (1ec5) [inline Tests/MapboxNavigationTests/StatusViewTests.swift:58]: There’s more to [adding a localizable string](https://github.com/mapbox/mapbox-navigation-ios/blob/main/CONTRIBUTING.md#adding-user-facing-text) to this codebase, but for an automated test, we actually want the string to be more predictable, so it would be better to avoid `NSLocalizedString` and hard-code `first test status` as is.

File: Sources/MapboxNavigation/StatusView.swift
Comment (1ec5) [reply Sources/MapboxNavigation/StatusView.swift:53]: Per https://github.com/mapbox/mapbox-navigation-ios/pull/2747#discussion_r583781129, it probably is necessary to distinguish between the ID and title, so we’d need to uncomment this property.

**Comment by 1ec5** [comment]
Looks like the PR branch got munged up somehow, as if it got merged then rebased. If `git reflog` says merging was the last thing you did, you could try hard-resetting to before that merge then performing just a rebase and finally force-pushing.

**Comment by jill-cardamon** [comment]
The last thing there is a rebase; I didn't use a merge at all.

**Comment by 1ec5** [comment]
I think this branch was rebased onto an old copy of the main branch.

File: CHANGELOG.md
Comment (1ec5) [inline CHANGELOG.md:6]: This change looks like a bad merge; can you revert it? Thanks!

File: Sources/MapboxNavigation/StatusView.swift
Comment (1ec5) [inline Sources/MapboxNavigation/StatusView.swift:61]: These properties also need to be public (and documented) in order for developers to set them.

File: Sources/MapboxNavigation/StatusView.swift
Comment (1ec5) [inline Sources/MapboxNavigation/StatusView.swift:79]: The documentation comment for Priority should indicate that lower values are higher priority.

File: Sources/MapboxNavigation/StatusView.swift
Comment (1ec5) [inline Sources/MapboxNavigation/StatusView.swift:68]: Since no specific values are defined, it would be simpler to define a type alias than a raw representable type alias inside a struct:

```
public typealias Priority = Int
```

File: Sources/MapboxNavigation/StatusView.swift
Comment (1ec5) [inline Sources/MapboxNavigation/StatusView.swift:195]: This should be a simple `==` check now that identifiers are independent of labels.

**Review by 1ec5 — COMMENTED** [review]
This is tantalizingly close – I think the remaining tasks will be pretty small in nature.

File: Sources/MapboxNavigation/StatusView.swift
Comment (1ec5) [inline Sources/MapboxNavigation/StatusView.swift:56]: FYI, it isn’t necessary to explicitly name the property in the property’s own documentation comment. For example, this could say “A string that uniquely identifies this status.”

File: Sources/MapboxNavigation/StatusView.swift
Comment (1ec5) [inline Sources/MapboxNavigation/StatusView.swift:95]: Replace the struct with the typealias. The `value` property will go away: you’ll be able to use an Int as if it were a Priority.

The previous version with the raw-representable Priority would’ve been appropriate if we wanted to define some preset values like `high` and `low` that were tied to specific values. But only a bare typealias is necessary if we’re treating it as just an ordinary number with special semantics.

**Review by 1ec5 — APPROVED** [review]
This is good to go once the struct goes away.

**Pull request merged**
